runtime/runtime-core/common/src/aws/smithy/kotlin/runtime/retries/RetryPolicy.kt

```diff
@@ -0,0 +1,18 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+package aws.smithy.kotlin.runtime.retries
+
+/**
+ * A policy that evaluates a [Result] from a retry attempt and indicates action a [RetryStrategy] should take next.
+ */
+interface RetryPolicy<in R> {
+    /**
+     * Evaluate the given retry attempt.
+     * @param result The [Result] of the retry attempt.
+     * @return A [RetryDirective] indicating what action the [RetryStrategy] should take next.
+     */
+    fun evaluate(result: Result<R>): RetryDirective
```

Comment:
question

I feel that this was brought up in discussion but I forget, what's the reasoning behind which functions suspend vs which do not?

Reply:
For retries, I put `suspend` around functions that would either need to execute the code block (which probably makes IO calls) or delay (e.g., backoff, token bucket interactions, etc.). All the retry policies that I've imagined so far all happen in straightforward, on-box logic. While I suppose it's technically possible some retry policy could involve network calls or local storage calls, it seems unlikely that we'll ever need that for the bare smithy or SDK use cases.